media_driver/agnostic/common/os/mos_sw_swizzle.c

```diff
@@ -0,0 +1,1029 @@
+/*
+* Copyright (c) 2019, Intel Corporation
+*
+* Permission is hereby granted, free of charge, to any person obtaining a
+* copy of this software and associated documentation files (the "Software"),
+* to deal in the Software without restriction, including without limitation
+* the rights to use, copy, modify, merge, publish, distribute, sublicense,
+* and/or sell copies of the Software, and to permit persons to whom the
+* Software is furnished to do so, subject to the following conditions:
+*
+* The above copyright notice and this permission notice shall be included
+* in all copies or substantial portions of the Software.
+*
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+* OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+* THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+* OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+* ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+* OTHER DEALINGS IN THE SOFTWARE.
+*/
+//!
+//! \file     mos_sw_swizzle.c
+//! \brief    Mesa ISL (Intel Surface Layout) implementation for sw-based tiling/detiling
+//! \details  Mesa ISL (Intel Surface Layout) implementation for sw-based tiling/detiling. SSE optimized
+//!
+
+#include <string.h>
+#include <immintrin.h>
+#include "mos_sw_swizzle.h"
+
+#define INLINE_SSE41
+#define USE_SSE41
+
+typedef void *(*isl_mem_copy_fn)(void *dest, const void *src, size_t n);
```

Comment:
Looks like Mesa has 3 files with memcpy:
* https://github.com/mesa3d/mesa/blob/master/src/intel/isl/isl_tiled_memcpy.c
* https://github.com/mesa3d/mesa/blob/master/src/intel/isl/isl_tiled_memcpy_normal.c
* https://github.com/mesa3d/mesa/blob/master/src/intel/isl/isl_tiled_memcpy_sse41.c

Which one you brought and why you did not brought others (if you did not)?

We need to make sure that this memcpy will work on all platforms we support. For example, check that we don't use sse4.2... And consider to have 2 paths: one for sse4.2, another for sse4.1...

Reply:
isl_tiled_memcpy_normal.c and isl_tiled_memcpy_sse41.c are just wrappers. real implementation is in isl_tiled_memcpy.c